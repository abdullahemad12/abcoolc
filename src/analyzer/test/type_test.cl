class Main inherits A
{
	meth1() : A { new SELF_TYPE };
	meth2() : SELF_TYPE {
		new SELF_TYPE
	};

	bin : Bin <- new SELF_TYPE;
	s : A <- new SELF_TYPE;
	a : A <- new A;
	m : A <- new B;
	b : B <- new A; 
	x : Int <- 0;
	main() : Int { hello1() };
	fun1() : Int { z };
	fun2() : String { x };
	fun3() : A { a };
	setB() : Main {
		{
			b <- new Bin;
			b <- new A;
			b <- new Bin;
			self;
		}
	};
	dispatch() : Int {
		new E.get_int()
	};
	
	dispatch2() : String {
		new E.get_int()
	};
	
};
class A inherits IO
{
	static_dispatch3() : Int {
		new B@E.parent_method(10, new Main, new Bin)
	};
	meth() : Main { new SELF_TYPE };
	y : Int <- 0;
	hello() : Int { y };
	bin_meth(x : Int, y : String, n : Main, i : IO, bin : Bin) : SELF_TYPE {
		self
	};
	bin_self_dispatch() : B {
		(new E).parent_method_self(10)
	};
	bin_self_dispatch1() : E {
		(new F).parent_method_self(10)
	};

	static_dispatch() : Int {
		(new B)@E.get_int()
	};

	static_dispatch1() : Int {
		(new B)@F.get_int()
	};

	static_dispatch2() : Int {
		(new A)@F.get_int()
	};


};

class Bin inherits B
{

};
class C inherits Main 
{
};

class D inherits C
{

};

class E inherits C
{
	get_int() : Int { 1 };
	get_self() : SELF_TYPE { self };
	my_method(x : Int) : SELF_TYPE { self };
	parent_method(x : Int, a : A, b : B) : Int
	{
		x
	};
	parent_method_self_err(x : Int, a : A) : SELF_TYPE
	{
		new F
	};

	parent_method_self(x : Int) : SELF_TYPE
	{
		self
	};

	parent_method_self1 (x : Int) : SELF_TYPE
	{
		new SELF_TYPE
	};

	self_self_dispatch() : SELF_TYPE
	{
		new E@SELF_TYPE.parent_method_self1(1).parent_method_self1(1)
	};
	
};

class F inherits E{
	my_method(x : Int) : SELF_TYPE { self };
};
class B inherits E {
	m : B <- new Main;
	my_method(x : Int, s : String) : SELF_TYPE
	{
		self
	};
	my_method1(m : String) : String
	{
		self.parent_method(1, new A, new Bin)
	};
	parent_method_self(x : Int) : Int {
		1
	};
};
