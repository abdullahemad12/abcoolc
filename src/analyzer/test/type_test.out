#1
_program
  #1
  _class
    Main
    A
    "test/type_test.cl"
    (
    #3
    _method
      meth1
      A
      #3
      _new
        SELF_TYPE
      : _no_type
    #4
    _method
      meth2
      SELF_TYPE
      #5
      _new
        SELF_TYPE
      : _no_type
    #8
    _attr
      bin
      Bin
      #8
      _new
        SELF_TYPE
      : _no_type
    #9
    _attr
      s
      A
      #9
      _new
        SELF_TYPE
      : _no_type
    #10
    _attr
      a
      A
      #10
      _new
        A
      : _no_type
    #11
    _attr
      m
      A
      #11
      _new
        B
      : _no_type
    #12
    _attr
      b
      B
      #12
      _new
        A
      : _no_type
    #13
    _attr
      x
      Int
      #13
      _int
        0
      : _no_type
    #14
    _method
      main
      Int
      #14
      _dispatch
        #14
        _object
          self
        : _no_type
        hello1
        (
        )
      : _no_type
    #15
    _method
      fun1
      Int
      #15
      _object
        z
      : _no_type
    #16
    _method
      fun2
      String
      #16
      _object
        x
      : _no_type
    #17
    _method
      fun3
      A
      #17
      _object
        a
      : _no_type
    #18
    _method
      setB
      Main
      #19
      _block
        #20
        _assign
          b
          #20
          _new
            Bin
          : _no_type
        : _no_type
        #21
        _assign
          b
          #21
          _new
            A
          : _no_type
        : _no_type
        #22
        _assign
          b
          #22
          _new
            Bin
          : _no_type
        : _no_type
        #23
        _object
          self
        : _no_type
      : _no_type
    #26
    _method
      dispatch
      Int
      #27
      _dispatch
        #27
        _new
          E
        : _no_type
        get_int
        (
        )
      : _no_type
    #30
    _method
      dispatch2
      String
      #31
      _dispatch
        #31
        _new
          E
        : _no_type
        get_int
        (
        )
      : _no_type
    )
  #35
  _class
    A
    IO
    "test/type_test.cl"
    (
    #37
    _method
      static_dispatch3
      Int
      #38
      _static_dispatch
        #38
        _new
          B
        : _no_type
        E
        parent_method
        (
        #38
        _int
          10
        : _no_type
        #38
        _new
          Main
        : _no_type
        #38
        _new
          Bin
        : _no_type
        )
      : _no_type
    #40
    _method
      meth
      Main
      #40
      _new
        SELF_TYPE
      : _no_type
    #41
    _attr
      y
      Int
      #41
      _int
        0
      : _no_type
    #42
    _method
      hello
      Int
      #42
      _object
        y
      : _no_type
    #43
    _method
      bin_meth
      #43
      _formal
        x
        Int
      #43
      _formal
        y
        String
      #43
      _formal
        n
        Main
      #43
      _formal
        i
        IO
      #43
      _formal
        bin
        Bin
      SELF_TYPE
      #44
      _object
        self
      : _no_type
    #46
    _method
      bin_self_dispatch
      B
      #47
      _dispatch
        #47
        _new
          E
        : _no_type
        parent_method_self
        (
        #47
        _int
          10
        : _no_type
        )
      : _no_type
    #49
    _method
      bin_self_dispatch1
      E
      #50
      _dispatch
        #50
        _new
          F
        : _no_type
        parent_method_self
        (
        #50
        _int
          10
        : _no_type
        )
      : _no_type
    #53
    _method
      static_dispatch
      Int
      #54
      _static_dispatch
        #54
        _new
          B
        : _no_type
        E
        get_int
        (
        )
      : _no_type
    #57
    _method
      static_dispatch1
      Int
      #58
      _static_dispatch
        #58
        _new
          B
        : _no_type
        F
        get_int
        (
        )
      : _no_type
    #61
    _method
      static_dispatch2
      Int
      #62
      _static_dispatch
        #62
        _new
          A
        : _no_type
        F
        get_int
        (
        )
      : _no_type
    )
  #68
  _class
    Bin
    B
    "test/type_test.cl"
    (
    )
  #72
  _class
    C
    Main
    "test/type_test.cl"
    (
    )
  #76
  _class
    D
    C
    "test/type_test.cl"
    (
    )
  #81
  _class
    E
    C
    "test/type_test.cl"
    (
    #83
    _method
      get_int
      Int
      #83
      _int
        1
      : _no_type
    #84
    _method
      get_self
      SELF_TYPE
      #84
      _object
        self
      : _no_type
    #85
    _method
      my_method
      #85
      _formal
        x
        Int
      SELF_TYPE
      #85
      _object
        self
      : _no_type
    #86
    _method
      parent_method
      #86
      _formal
        x
        Int
      #86
      _formal
        a
        A
      #86
      _formal
        b
        B
      Int
      #88
      _object
        x
      : _no_type
    #90
    _method
      parent_method_self_err
      #90
      _formal
        x
        Int
      #90
      _formal
        a
        A
      SELF_TYPE
      #92
      _new
        F
      : _no_type
    #95
    _method
      parent_method_self
      #95
      _formal
        x
        Int
      SELF_TYPE
      #97
      _object
        self
      : _no_type
    #100
    _method
      parent_method_self1
      #100
      _formal
        x
        Int
      SELF_TYPE
      #102
      _new
        SELF_TYPE
      : _no_type
    #105
    _method
      self_self_dispatch
      SELF_TYPE
      #107
      _dispatch
        #107
        _static_dispatch
          #107
          _new
            E
          : _no_type
          SELF_TYPE
          parent_method_self1
          (
          #107
          _int
            1
          : _no_type
          )
        : _no_type
        parent_method_self1
        (
        #107
        _int
          1
        : _no_type
        )
      : _no_type
    )
  #112
  _class
    F
    E
    "test/type_test.cl"
    (
    #113
    _method
      my_method
      #113
      _formal
        x
        Int
      SELF_TYPE
      #113
      _object
        self
      : _no_type
    )
  #115
  _class
    B
    E
    "test/type_test.cl"
    (
    #116
    _attr
      m
      B
      #116
      _new
        Main
      : _no_type
    #117
    _method
      my_method
      #117
      _formal
        x
        Int
      #117
      _formal
        s
        String
      SELF_TYPE
      #119
      _object
        self
      : _no_type
    #121
    _method
      my_method1
      #121
      _formal
        m
        String
      String
      #123
      _dispatch
        #123
        _object
          self
        : _no_type
        parent_method
        (
        #123
        _int
          1
        : _no_type
        #123
        _new
          A
        : _no_type
        #123
        _new
          Bin
        : _no_type
        )
      : _no_type
    #125
    _method
      parent_method_self
      #125
      _formal
        x
        Int
      Int
      #126
      _int
        1
      : _no_type
    )
